[EDU-3829] feat: add manage objects in buckets guide (#626)
aziontech/docs: src/content/docs/pt-br/pages/guias/object-storage/bucket-actions/index.mdx

---
title: Como fazer o upload e download de objetos de um bucket do Azion Object Storage
description: Aprenda como fazer upload e download de arquivos de buckets do Azion Object Storage.
meta_tags: object storage, storage, cloud, s3, bucket, file storage
namespace: documentation_products_object_storage_upload

> permalink: /documentacao/produtos/guias/fazer-upload-e-download-de-objetos-do-bucket/

Reviewer: ```suggestion
permalink: /documentacao/produtos/guias/upload-e-download-de-objetos-do-bucket/
```

> permalink: /documentacao/produtos/guias/como-fazer-upload-e-download-de-objetos-do-bucket/

Author: ```suggestion
permalink: /documentacao/produtos/guias/fazer-upload-e-download-de-objetos-do-bucket/
```

---

import Button from '~/components/Button.astro'

> **Object Storage** permite que você crie buckets e integre seu armazenamento de objetos com sua infraestrutura do edge. Este guia abrange como fazer upload e download de objetos de um bucket usando a [API da Azion](https://api.azion.com) e [Edge Runtime](/pt-br/documentation/devtools/runtime/overview/).

Reviewer: ```suggestion
**Azion Object Storage** permite que você crie buckets e integre seu armazenamento de objetos com sua infraestrutura do edge. Este guia abrange como fazer upload e download de objetos de um bucket usando a [API da Azion](https://api.azion.com) e [Edge Runtime](/pt-br/documentation/devtools/runtime/overview/).
```

<Button href="/pt-br/documentacao/produtos/store/object-storage/" text="vá para a documentação de referência de Object Storage" variant="secondary"></Button>

> Consulte o guia [Como criar e modificar um bucket de Object Storage da Azion](/pt-br/documentacao/produtos/guias/como-criar-e-modificar-um-bucket/) para operações em buckets.

Reviewer: ```suggestion
Consulte o guia [Como criar e modificar um bucket de Object Storage da Azion](/pt-br/documentacao/produtos/guias/criar-e-modificar-um-bucket/) para operações em buckets.
```

---

## Via API

> ### Fazer upload de objeto para um bucket

Reviewer: ```suggestion
### Faça upload de objeto para um bucket
```

Para fazer o upload de um objeto, execute a seguinte requisição `POST` em seu terminal, substituindo `[TOKEN VALUE]` por seu [personal token](/pt-br/documentacao/produtos/guias/personal-tokens/), `<bucket_name>` pelo nome do seu bucket, `<object_key>` por um ID ou nome para o objeto, o cabeçalho `Content-Type` com o MIME type sendo enviado, e o arquivo enviado convertido em data binary:

```bash
curl --location 'https://api.azion.com/v4/storage/buckets/<bucket_name>/objects/<object_key>' \
--header 'Accept: application/json; version=3' \
--header 'Content-Type: text/csv' \
--header 'Authorization: Token [TOKEN VALUE]' \
--data '@./path/file.csv'
```

Você deverá receber a seguinte resposta:

```json
{
    "state": "executed",
    "data": {
        "object_key": "folder/csv-file"
    }
}
```

> ## Fazer download de objeto de um bucket

Reviewer: ```suggestion
## Faça download de objeto de um bucket
```

Para fazer o download de um objeto, execute a seguinte requisição `GET` em seu terminal, substituindo `[TOKEN VALUE]` por seu [personal token](/pt-br/documentacao/produtos/guias/personal-tokens/), `<bucket_name>` pelo nome do seu bucket, e `<object_key>` pela chave criada para o objeto:

```bash
curl --location 'https://api.azion.com/v4/storage/buckets/<bucket_name>/objects/<object_key>' \
--header 'Accept: application/json; version=3' \
--header 'Authorization: Token [TOKEN VALUE]'
```

---

## Via Edge Runtime

Você pode criar uma edge function para fazer upload de arquivos para seu bucket usando um comando de console `POST` e verificar se um arquivo está disponível usando um comando de console `GET`. Para fazer isso:

> 1. Accesse o [Real-Time Manager (RTM)](https://manager.azion.com/) > **Edge Functions**.

Reviewer 2: ```suggestion
1. Accesse o [Real-Time Manager (RTM)](/pt-br/documentacao/produtos/guias/como-acessar-o-rtm/) > **Edge Functions**.
```

2. Clique no botão **Add Function**.

> 3. Dê um nome a sua função. Exemplo: `my-bucket POST+GET`

Reviewer: ```suggestion
3. Dê um nome a sua função. Exemplo: `my-bucket POST+GET`.
```

> 4. Na aba **Code**, adicione o seguinte código Javascript: 

Reviewer: ```suggestion
4. Na aba **Code**, adicione o seguinte código JavaScript:
```

    return new Response(await asset.arrayBuffer(), {
        headers: {
            "Content-Type": asset.contentType
        },
    });
}

async function doPost(path, content_type, value, bucket_name) {
    console.log(`Path: ${path}`);
    let options = {
        "Content-Type": content_type
    }
    console.log(`Options: ${options}`);
    const storage = new Storage(bucket_name);
    await storage.put(path, value, options);
    return new Response("Object added.");
}

async function router(event) {
    const request = event.request;
    const method = request.method;
    const path = decodeURI(new URL(request.url).pathname);
    const bucket_name = event.args.bucket;
    if (method === "POST") {
        let content_type = request.headers.get("Content-Type");
        let content = await request.arrayBuffer();
        return doPost(path, content_type, content, bucket_name);
    } else if (method === "GET") {
        return doGet(path, bucket_name);
    } else {
        throw new Error(`Invalid method: ${method}. Expected POST or GET.`);
    }
}
addEventListener("fetch", (event) => {
    event.respondWith(
        router(event)
    );
});
```

> | Variável | Description |

Reviewer: ```suggestion
| Variável | Descrição |
```

> | Variável | Valor esperado |

Author: ```suggestion
| Variável | Description |
```

> | Variável | Valor esperado |

Reviewer: mesmo comentário do label que fiz em EN

| --- | --- |
| `path` | Caminho do objeto. Exemplo: `./path/file.csv` |
| `bucket_name` | Nome do bucket. Exemplo: `my-bucket` |
| `content_type` | MIME type do objeto. Exemplo: `text/csv` |
| `value` | Conteúdo do objeto em data binary |

5. Na aba **Arguments**, adicione o objeto com a propriedade `bucket` e, como valor, o nome do bucket em string.

```json
{
  "bucket": "my-bucket"
}
```

6. Clique no botão **Save**.

:::note

> Consulte a referência [Object Storage API](/en/documentation/devtools/runtime/api-reference/storage/) para mais detalhes nas funções disponíveis para outros tipos de requisições.

Reviewer: ```suggestion
Consulte a referência [Object Storage API](/en/documentation/devtools/runtime/api-reference/storage/) para mais detalhes sobre as funções disponíveis para outros tipos de requisições.
```

:::

> Após criar a edge function, você precisa criar uma edge application que vai intermediar o processo de upload para o Object Storage.

Author: ```suggestion
Após criar a edge function, você precisa criar uma edge application que vai intermediar o processo de upload para o bucket.
```

> Uma vez que você tenha a edge function, você precisa criar uma edge application que vai intermediar o processo de upload para o Object Storage.

Reviewer: ```suggestion
Após criar a edge function, você precisa criar uma edge application que vai intermediar o processo de upload para o Object Storage.
```

> 1. Acesse o [Real-Time Manager (RTM)](https://manager.azion.com/)

Reviewer: ```suggestion
1. Acesse o [Real-Time Manager (RTM)](https://manager.azion.com/).
```
